### PR TITLE
Implement SQL Analysis worker

### LINKED ISSUE #45462: Implement SQL Analyzer worker
Add an always-running background thread, which reads Card ids off a queue, and executes `query-field/update-query-fields-for-card!`.

While we are here, move this mutating method to a new namespace which can group the other kinds of models we will use for result columns and source tables.

We also want to introduce a feature flag to skip analysis for MBQL queries. Move this feature check up a level into `query-field-ids` as well, and clean up the tests.
```[tasklist]
### Tasks
- [ ] https://github.com/metabase/metabase/pull/45483
- [ ] https://github.com/metabase/metabase/pull/45543
- [ ] https://github.com/metabase/metabase/pull/45549
```

### LINKED ISSUE #45463: Update Card modification hooks to use SQL Analysis queue
Instead of directly calling `update-query-fields-for-card!` in the `after-insert` and `after-update` hooks, just enqueue it. 

This enqueue method should probably we smart and skip enqueuing if the relevant feature flag is disabled. We can probably remove checking the flag in the Analyzer then.

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes https://github.com/metabase/metabase/issues/45462
Closes https://github.com/metabase/metabase/issues/45463

### Description

With this change, all query analysis happens off the main thread, and it is rate limited in order to run less than 20% of the time averaged out over time.

### CODE REVIEW

File: test/metabase/task/analyze_queries_test.clj [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+   [metabase.util.queue :as queue]
+   [toucan2.core :as t2]))
+
+(deftest analyzer-loop-test
+  (let [metadata-provider (lib.metadata.jvm/application-database-metadata-provider (mt/id))
+        venues            (lib.metadata/table metadata-provider (mt/id :venues))
+        venues-name       (lib.metadata/field metadata-provider (mt/id :venues :name))
+        mlv2-query        (-> (lib/query metadata-provider venues)
+                              (lib/aggregate (lib/distinct venues-name)))]
+    (mt/with-temp [Card c1   {:query_type    "native"
```
Comment (crisptrutski) [inline test/metabase/task/analyze_queries_test.clj:21]: It'd be nice to DRY these fixtures up with the backfill / sweeper test as well. Will probably do that when I rework that task.

**Comment by crisptrutski** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/45549?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#45551** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45551?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#45549** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45549?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#45543** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45543?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#45483** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45483?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @crisptrutski and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: .clj-kondo/config.edn [hunk 2]
```diff
@@ -794,7 +794,7 @@
    metabase-enterprise.serialization.test-util/with-world                       macros.metabase-enterprise.serialization.test-util/with-world
    metabase-enterprise.test/with-gtaps!                                         macros.metabase-enterprise.sandbox.test-util/with-gtaps!
    metabase-enterprise.advanced-permissions.api.util-test/with-impersonations!  macros.metabase-enterprise.advanced-permissions.api.util-test/with-impersonations!
-   metabase-enterprise.query-field-validation.api-test/with-test-setup          macros.metabase-enterprise.query-field-validation.api-test/with-test-setup
+   metabase-enterprise.query-field-validation.api-test/with-test-setup!         macros.metabase-enterprise.query-field-validation.api-test/with-test-setup!
```
Comment (crisptrutski) [inline .clj-kondo/config.edn:797]: Can roll this back. Kondo made me do it for an earlier version where I was calling bang method in the implementation.

File: src/metabase/task/backfill_query_fields.clj [hunk 3]
```diff
@@ -10,19 +10,21 @@
 
 (set! *warn-on-reflection* true)
 
-(defn- backfill-query-fields! []
-  (let [cards (t2/reducible-select :model/Card :id [:in {:from      [[:report_card :c]]
```
Comment (crisptrutski) [inline src/metabase/task/backfill_query_fields.clj:14]: The semantics of this query are the same, except it no longer uses a subselect. This means it is both more efficient, and (more importantly) easier to extend as the sweeper gets smarter.

File: src/metabase/task/backfill_query_fields.clj [hunk 4]
```diff
@@ -10,19 +10,21 @@
...
+                                     :where     [:and
+                                                 [:not :report_card.archived]
+                                                 [:= :f.id nil]]})]
+     (run! analyze-fn cards))))
 
 (jobs/defjob ^{DisallowConcurrentExecution true
                :doc                        "Backfill QueryField for cards created earlier. Runs once per instance."}
              BackfillQueryField [_ctx]
-  (backfill-query-fields!))
+  (backfill-missing-query-fields!))
```
Comment (crisptrutski) [inline src/metabase/task/backfill_query_fields.clj:27]: I distinguish "missing" here, as in future we'll also retry stale, etc.

File: src/metabase/config.clj [hunk 5]
```diff
@@ -84,9 +84,13 @@
 (defn config-bool "Fetch a configuration key and parse it as a boolean."  ^Boolean [k] (some-> k config-str Boolean/parseBoolean))
 (defn config-kw   "Fetch a configuration key and parse it as a keyword."  ^Keyword [k] (some-> k config-str keyword))
 
-(def ^Boolean is-dev?  "Are we running in `dev` mode (i.e. in a REPL or via `clojure -M:run`)?" (= :dev  (config-kw :mb-run-mode)))
-(def ^Boolean is-prod? "Are we running in `prod` mode (i.e. from a JAR)?"                       (= :prod (config-kw :mb-run-mode)))
-(def ^Boolean is-test? "Are we running in `test` mode (i.e. via `clojure -X:test`)?"            (= :test (config-kw :mb-run-mode)))
+(def run-mode
```
Comment (tsmacdonald) [inline src/metabase/config.clj:87]:
```suggestion
(def ^:private run-mode
```

A big benefit of `is-{dev/prod/test}?` is that we don't have to worry about whether the run mode is spelled `:dev` or `:development`, etc. It'd be nice to keep the specific keywords used locked up in this ns

File: src/metabase/query_analysis.clj [hunk 6]
```diff
@@ -12,29 +12,59 @@
...
+(defmacro without-analysis
+  "Override the default execution mode to always use the current thread. Does nothing in prod - only use this in tests."
+  [& body]
+  `(with-execution* ::disabled ~@body))
+
+(defn- execution []
+  (case config/run-mode
+    :prod ::queued
+    :dev  *analyze-execution-in-dev?*
+    :test *analyze-execution-in-test?*))
```
Comment (tsmacdonald) [inline src/metabase/query_analysis.clj:60]: Ah, I see why it's public. idk, I might've written a `cond` with `is-dev?` and friends, but maybe I'm overthinking it. This is not a blocking comment.

File: src/metabase/query_analysis.clj [hunk 7]
```diff
@@ -12,29 +12,59 @@
...
 
-  c.f. [[*parse-queries-in-test?*]], [[public-settings/sql-parsing-enabled]]"
-  []
-  (and (public-settings/sql-parsing-enabled)
-       (or (not config/is-test?)
-           *parse-queries-in-test?*)))
 
-(defn enabled?
+(defn enabled-type?
```
Comment (tsmacdonald) [inline src/metabase/query_analysis.clj:63]: `query-type-enabled?`?

File: src/metabase/query_analysis.clj (hunk 6 above)
Comment (crisptrutski) [reply src/metabase/query_analysis.clj:60]: If it's controversial the `cond` route is definitely the next best.

File: src/metabase/query_analysis.clj (hunk 7 above)
Comment (crisptrutski) [reply src/metabase/query_analysis.clj:63]: That sounds it bit yoda to me "query analysis query type enabled?"

If we want to be more explicit how about `query-analysis/enabled-for-query-type?`?

File: src/metabase/task/analyze_queries.clj [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+   [metabase.util.log :as log])
+  (:import
+   (org.quartz DisallowConcurrentExecution)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private max-cpu-usage-ratio 0.2)
+
+(def ^:private wait-ratio
+  (/ (- 1 max-cpu-usage-ratio) max-cpu-usage-ratio))
```
Comment (tsmacdonald) [inline src/metabase/task/analyze_queries.clj:19]: For the benefit of any bozos looking at this:

image

Comment (tsmacdonald) [reply src/metabase/task/analyze_queries.clj:19]: (it's me. I'm the bozo)

File: src/metabase/task/analyze_queries.clj [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+             (analyzer-loop!))
+
+(defmethod task/init! ::BackfillQueryField [_]
+  (let [job     (jobs/build
+                 (jobs/of-type QueryAnalyzer)
+                 (jobs/with-identity (jobs/key "metabase.task.analyze-queries.job")))
+        trigger (triggers/build
+                 (triggers/with-identity (triggers/key "metabase.task.analyze-queries.trigger"))
+                 (triggers/with-schedule
+                  (simple/schedule (simple/with-interval-in-minutes 1)))
```
[the remark this answers is not in the document]
Comment (tsmacdonald) [reply src/metabase/task/analyze_queries.clj:59]: :grimacing: I'd ask in #engineering-backend and tag Cam

**Review by tsmacdonald — APPROVED** [review]
Seems good to me :+1:

none of my comments are important

File: src/metabase/task/analyze_queries.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (crisptrutski) [reply src/metabase/task/analyze_queries.clj:59]: Not going to block on this, as we will come back to it during manual and e2e testing.

File: .clj-kondo/config.edn (hunk 2 above)
Comment (crisptrutski) [reply .clj-kondo/config.edn]: Rolled it back

File: src/metabase/task/analyze_queries.clj (hunk 8 above)
Comment (crisptrutski) [reply src/metabase/task/analyze_queries.clj:19]: Thanks! Maybe we could capture the math in a comment or unit test.

It feels weird testing an equation though.

**Pull request merged**